File: Python3/src/fssearch/format_number.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
 * 
 *  Coded by Rei-Chi Lin 
 * 
"""

from decimal import Decimal

__author = "Rei-Chi Lin"
def author():
    return __author

def padding_number_to_3_digits(number):
    if number < 0:
        return '-' + padding_number_to_3_digits(abs(number))
    numeric_str = ''
    if number == 0:
        numeric_str = '000'
    else:
        if number < 10:
            numeric_str = '00' + str(number)
        else:
            if number < 100:
                numeric_str = '0' + str(number)
            else:
                numeric_str = str(number)
    return numeric_str

def format_number_kilo_by_kilo(number):
    if number == 0:
        return "0"
    elif number < 0:
        return "-" + format_number_kilo_by_kilo(abs(number))
    elif int(number) != number:
        return format_number_kilo_by_kilo(int(number)) + str(number - int(number))[1:]
    
    formatted_str = ""
    count = 0
    while number > 0:
        if number // 1000 == 0:
            if count > 0:
                formatted_str = str(number % 1000) + "," + formatted_str
            else:
                formatted_str = str(number % 1000) + formatted_str
        else:
            if count > 0:
                formatted_str = padding_number_to_3_digits(number % 1000) + "," + formatted_str
            else:
                formatted_str = padding_number_to_3_digits(number % 1000) + formatted_str
        number = number // 1000
        count += 1
    return formatted_str

if __name__ == "__main__":
    # for unit test
    num = Decimal(-1002073000.75)
    print("input:\n\t" + str(num))
    print("output:\n\t" + format_number_kilo_by_kilo(num))

File: Rust/src/format_number.rs
pub fn padding_number_to_3_digits(number: i32) -> String {
    if number < 0 {
        return format!("-{}", padding_number_to_3_digits(number.abs()));
    }
    let numeric_str = if number == 0 {
        "000".to_string()
    } else if number < 10 {
        format!("00{}", number)
    } else if number < 100 {
        format!("0{}", number)
    } else {
        number.to_string()
    };
    numeric_str
}

pub fn format_number_kilo_by_kilo(number: i64) -> String {
    if number == 0 {
        return "0".to_string();
    } else if number < 0 {
        return format!("-{}", format_number_kilo_by_kilo(-number));
    }
    
    let mut number = number;
    let mut formatted_str = String::new();
    let mut count = 0;
    
    while number > 0 {
        let chunk = number % 1000;
        if number >= 1000 {
            formatted_str = format!(",{:03}", chunk) + &formatted_str;
        } else {
            formatted_str = chunk.to_string() + &formatted_str;
        }
        number /= 1000;
        count += 1;
    }
    
    formatted_str
}